framework/src/geomsearch/SlaveNeighborhoodThread.C

```diff
@@ -70,40 +62,30 @@ SlaveNeighborhoodThread::operator()(const NodeIdRange & range)
 {
   processor_id_type processor_id = _mesh.processor_id();
 
-  for (const auto & node_id : range)
-  {
-    const Node & node = *_mesh.nodePtr(node_id);
-
-    std::priority_queue<std::pair<unsigned int, Real>,
-                        std::vector<std::pair<unsigned int, Real>>,
-                        ComparePair>
-        neighbors;
+  unsigned int patch_size = std::min(
+      _patch_size, static_cast<unsigned int>(_trial_master_nodes.size()));
 
-    unsigned int n_master_nodes = _trial_master_nodes.size();
+  std::vector<size_t> ret_index(patch_size);
+  std::vector<Real> ret_dist_sqr(patch_size);
 
-    // Get a list, in descending order of distance, of master nodes in relation to this node
-    for (unsigned int k = 0; k < n_master_nodes; k++)
-    {
-      dof_id_type master_id = _trial_master_nodes[k];
-      const Node * cur_node = _mesh.nodePtr(master_id);
-      Real distance = ((*cur_node) - node).norm();
-
-      neighbors.push(std::make_pair(master_id, distance));
+  for (const auto &node_id : range) {
+    for (unsigned int i = 0; i < patch_size; ++i) {
+      ret_dist_sqr[i] = std::numeric_limits<Real>::max();
+      ret_index[i] = std::numeric_limits<size_t>::max();
     }
 
-    std::vector<dof_id_type> neighbor_nodes;
+    const Node &node = _mesh.nodeRef(node_id);
+    const Real query_pt[] = {node(0), node(1), node(2)};
```

Comment:
I am leaning towards the `KDTree` class. If I did not miss something, it can live outside `NearestNodeLocator` and I would say, put it in `utils/`. 

Also comments like "Need to convert this into a C-array because that's what nanoflann expects" will help explain things like this to people not familiar with the interface.